src/Functions/array/l2DistanceTransposed.cpp

```diff
@@ -179,14 +192,34 @@ class FunctionArrayDistance : public IFunction
         const auto qbit_size_bytes = DataTypeQBit::bitsToBytes(qbit_size);
 
         /// All QBit subcolumns should be FixedString and have a consistent size
-        for (size_t i = 0; i < arguments.size() - 2; ++i)
+        for (size_t i = 0; i < precision; ++i)
         {
             const auto * arg_type = checkAndGetDataType<DataTypeFixedString>(arguments[i].type.get());
 
             if (!arg_type || arg_type->getN() != qbit_size_bytes)
                 return false;
         }
 
+        /// The type of reference vector dictates what type QBit had before we sliced it into q.1, ..., q.precision.
+        /// Check that the number of bit planes doesn't exceed the maximum precision for the reference vector type.
+        size_t max_precision_for_type = 0;
+        if (ref_vec_type_id == TypeIndex::BFloat16)
+            max_precision_for_type = 16;
+        else if (ref_vec_type_id == TypeIndex::Float32)
+            max_precision_for_type = 32;
+        else if (ref_vec_type_id == TypeIndex::Float64)
+            max_precision_for_type = 64;
+        else
+            return false;
+
+        if (precision > max_precision_for_type)
+            throw Exception(
+                ErrorCodes::BAD_ARGUMENTS,
+                "Number of bit planes ({}) exceeds the maximum precision ({}) for reference vector type {}",
+                precision,
+                max_precision_for_type,
+                ref_vec_type->getNestedType()->getName());
```

Comment:
In other cases here we return `false` on any mismatch to fallback to non-optimized version, should we do the same here?